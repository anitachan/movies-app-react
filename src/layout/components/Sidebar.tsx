import { CSSObject } from '@emotion/react';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import FavoriteIcon from '@mui/icons-material/Favorite';
import HomeIcon from '@mui/icons-material/Home';
import {
  Divider,
  IconButton,
  List,
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
  Drawer as MuiDrawer,
  useTheme,
} from '@mui/material';
import { styled } from '@mui/material/styles';
import { Link } from 'react-router-dom';

interface Props {
  open: boolean;
  handleDrawerClose: () => void;
  drawerWidth: number;
}

const openedMixin = (theme: any, drawerWidth: number): CSSObject => ({
  width: drawerWidth,
  transition: theme.transitions.create('width', {
    easing: theme.transitions.easing.sharp,
    duration: theme.transitions.duration.enteringScreen,
  }),
  overflowX: 'hidden',
});

const closedMixin = (theme: any): CSSObject => ({
  transition: theme.transitions.create('width', {
    easing: theme.transitions.easing.sharp,
    duration: theme.transitions.duration.leavingScreen,
  }),
  overflowX: 'hidden',
  width: `calc(${theme.spacing(7)} + 1px)`,
  [theme.breakpoints.up('sm')]: {
    width: `calc(${theme.spacing(7)} + 1px)`,
  },
});

const DrawerHeader = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'flex-end',
  padding: theme.spacing(0, 1),
  ...theme.mixins.toolbar,
}));

const Drawer = styled(MuiDrawer, {
  shouldForwardProp: (prop) => prop !== 'open' && prop !== 'drawerWidth',
})<{ open: boolean; drawerWidth: number }>(({ theme, open, drawerWidth }) => ({
  width: drawerWidth,
  flexShrink: 0,
  whiteSpace: 'nowrap',
  position: 'relative',
  boxSizing: 'border-box',
  ...(open
    ? {
        ...openedMixin(theme, drawerWidth),
        '& .MuiDrawer-paper': openedMixin(theme, drawerWidth),
      }
    : {
        ...closedMixin(theme),
        '& .MuiDrawer-paper': closedMixin(theme),
      }),
}));

export const Sidebar = ({ open, handleDrawerClose, drawerWidth }: Props) => {
  const theme = useTheme();

  const menuItems = [
    { id: 'Home', icon: HomeIcon, route: '/home' },
    { id: 'Favorites', icon: FavoriteIcon, route: '/favorites' },
  ];

  return (
    <Drawer variant="permanent" open={open} drawerWidth={drawerWidth}>
      <DrawerHeader>
        <IconButton onClick={handleDrawerClose}>
          {theme.direction === 'ltr' ? <ChevronLeftIcon /> : <ChevronRightIcon />}
        </IconButton>
      </DrawerHeader>
      <Divider />
      <List>
        {menuItems.map(({ id, icon: IconComponent, route }) => (
          <ListItem key={id} disablePadding>
            <ListItemButton component={Link} to={route}>
              <ListItemIcon>
                <IconComponent />
              </ListItemIcon>
              <ListItemText primary={id} />
            </ListItemButton>
          </ListItem>
        ))}
      </List>
    </Drawer>
  );
};
