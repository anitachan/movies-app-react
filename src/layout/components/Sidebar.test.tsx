import { render, screen, fireEvent } from '@testing-library/react';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { Sidebar } from './Sidebar';

jest.mock(
  'react-router-dom',
  () => ({
    Link: ({ to, children }: any) => <a href={to}>{children}</a>,
  }),
  { virtual: true },
);

describe('Sidebar component', () => {
  const drawerWidth = 240;
  const theme = createTheme();

  const renderSidebar = (props: Partial<React.ComponentProps<typeof Sidebar>> = {}) =>
    render(
      <ThemeProvider theme={theme}>
        <Sidebar
          open={props.open ?? true}
          drawerWidth={props.drawerWidth ?? drawerWidth}
          handleDrawerClose={props.handleDrawerClose ?? jest.fn()}
        />
      </ThemeProvider>,
    );

  test('should show home and favorites and its routes', () => {
    renderSidebar();
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByText('Favorites')).toBeInTheDocument();

    const links = screen.getAllByRole('link');
    expect(links.map((a) => (a as HTMLAnchorElement).href)).toEqual([
      expect.stringContaining('/home'),
      expect.stringContaining('/favorites'),
    ]);
  });

  test('should call handleDrawerClose when IconButton is clicked', () => {
    const onClose = jest.fn();
    renderSidebar({ handleDrawerClose: onClose });
    fireEvent.click(screen.getByRole('button'));
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test('should set drawerWidth when open is true', () => {
    const { container } = renderSidebar({ open: true });
    const drawerRoot = container.firstChild as HTMLElement;
    expect(drawerRoot).toHaveStyle(`width: ${drawerWidth}px`);
  });

  test('should set closedMixin when open is false', () => {
    const { container } = renderSidebar({ open: false });
    const drawerRoot = container.firstChild as HTMLElement;
    expect(drawerRoot).toHaveStyle('width: calc(56px + 1px)');
  });

  test('should use ChevronRightIcon when theme.direction is rtl', () => {
    const rtlTheme = createTheme({ direction: 'rtl' });
    const { container } = render(
      <ThemeProvider theme={rtlTheme}>
        <Sidebar open={true} drawerWidth={240} handleDrawerClose={() => {}} />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
