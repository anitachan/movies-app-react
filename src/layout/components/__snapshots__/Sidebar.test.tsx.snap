// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should use ChevronRightIcon when theme.direction is rtl 1`] = `
<div>
  <div
    class="MuiDrawer-root MuiDrawer-anchorLeft MuiDrawer-docked css-1hg0i2n-MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-1lwhjos-MuiPaper-root-MuiDrawer-paper"
      style="--Paper-shadow: none;"
    >
      <div
        class="css-1r9jet7"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </button>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
      />
      <ul
        class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
        >
          <a
            href="/home"
          >
            <div
              class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="HomeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                />
              </svg>
            </div>
            <div
              class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Home
              </span>
            </div>
          </a>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
        >
          <a
            href="/favorites"
          >
            <div
              class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="FavoriteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
                />
              </svg>
            </div>
            <div
              class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Favorites
              </span>
            </div>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
